crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_types::committee::EpochId;
+use typed_store::Map;
+
+use std::sync::Arc;
+use sui_types::base_types::ObjectDigest;
+
+use fastcrypto::hash::{Digest, MultisetHash};
+use sui_types::accumulator::Accumulator;
+use sui_types::error::SuiResult;
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use typed_store::rocks::TypedStoreError;
+
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::authority::AuthorityStore;
+
+#[derive(Debug, Clone)]
+pub struct State {
+    pub effects: Vec<TransactionEffects>,
+    pub checkpoint_seq_num: CheckpointSequenceNumber,
+}
+
+pub struct StateAccumulator {
+    authority_store: Arc<AuthorityStore>,
+}
+
+impl StateAccumulator {
+    pub fn new(authority_store: Arc<AuthorityStore>) -> Self {
+        Self { authority_store }
+    }
+
+    pub fn accumulate_checkpoint(
+        &self,
+        state: State,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> SuiResult<Accumulator> {
+        if let Some(acc) = epoch_store.get_state_hash_for_checkpoint(&state.checkpoint_seq_num)? {
+            return Ok(acc);
+        }
+
+        let mut acc = Accumulator::default();
+
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.created
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
```

Comment:
Yes, the object digest uniquely identifies the object id + sequence number. To your question of the purpose of this pr, `StateAccumulator` ultimately aims to generate a root state hash that uniquely identifies the state of the system at end of epoch, as defined by the live object set (the set of object id + sequence number that currently exist in the system). Since we don't want to have to scan the live object set in its entirety at the end of the epoch while the network is halted, we instead process the transaction effects certificates from the checkpoint, and accumulate the incremental changes from genesis, which should result in the same end state. 

This root state hash is then sent through consensus and included in the `VerifiedCheckpoint` cert. 

This is part of a larger system that will generate snapshots of the live object set, which a new node to the network can quickly download, then hash, and verify that the hash that it arrives on matches the one in the VerifiedCheckpoint cert. Hope that helps!